### PR TITLE
Reader: add tracking for scroll depth on full post article

### PR DESCRIPTION
**Pull request by eoigal**
Fixes https://github.com/Automattic/loop/issues/262
Fixes https://github.com/Automattic/loop/issues/263

## Proposed Changes

1. Added new state variables `maxScrollDepth` and `hasCompleted` to track scroll depth and completion status.
2. Added an event listener for the 'scroll' event to track scroll depth.
3. Added methods `setScrollDepth`, `trackScrollDepth`, and `trackExitBeforeCompletion` to handle tracking scroll depth and user exit before completion events.
4. Updated `componentDidUpdate` to include calls to `trackScrollDepth` and `trackExitBeforeCompletion` when the post being viewed changes.
5. Updated `componentWillUnmount` to remove the scroll event listener.
6. Updated `handleVisibilityChange` to include calls to `trackScrollDepth` and `trackExitBeforeCompletion` when the document becomes hidden.
7. Updated `trackReadingTime` to check for post.ID.
8. Added logic to calculate and record the maximum scroll depth achieved and whether the user has completed the post.

## Why are these changes being made?

* Extra metrics to measure user engagement

## Testing Instructions

*  Go to `/read` and navigate between full post views
*  When you navigate away from the post, you should see the track event `calypso_reader_article_scroll_depth` logged
*  If the scroll depth doesn't go past 90% then a track event `calypso_reader_article_exit_before_completion` should also be logged.
*  You should be able to see these events on tracks after 5 mins
*  To see them live, I suggest adding console.log to the `recordTrackForPost` method.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/blocks/reader-full-post/index.jsx [hunk 1]
```diff
@@ -248,6 +264,48 @@ export class FullPostView extends Component {
 		}
 	}
 
+	setScrollDepth = () => {
+		if ( this.scrollableContainer ) {
+			const scrollTop = this.scrollableContainer.scrollTop;
```
Comment (spsiddarthan) [inline client/blocks/reader-full-post/index.jsx:271]: Are you able to get this block of code hit? No matter how scrollable my posts are, this bit of code never seems to get hit. I am attaching a gif as an example of what I am doing. 

CleanShot 2024-12-13 at 19 14 33

**Review by spsiddarthan — COMMENTED** [review]
The events are getting logged, but the scroll_depth parm is zero for all events. I maybe wrong, but shouldn't the scroll_depth be a non zero value for posts that can be scrolled?

Possible I am testing this wrong, so I will look forward to your views. :) 

CleanShot 2024-12-13 at 19 14 33

File: client/blocks/reader-full-post/index.jsx (hunk 1 above)
Comment (eoigal) [reply client/blocks/reader-full-post/index.jsx:271]: Scroll container is different between dataview and stream views, so needed to cater for that.

**Review by spsiddarthan — APPROVED** [review]
Works well now 🚢

**Pull request merged**